src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -0,0 +1,535 @@
+                }
+            }
+        }
+
+        /// <devdoc>
+        ///   <para>
+        ///     Determines if an address is an IPv6 Multicast address
+        ///   </para>
+        /// </devdoc>
+        public bool IsIPv6Multicast
+        {
+            get
+            {
+                return (_family == AddressFamily.InterNetworkV6) &&
+                    ((_numbers[0] & 0xFF00) == 0xFF00);
+            }
+        }
+
+        /// <devdoc>
+        ///   <para>
+        ///     Determines if an address is an IPv6 Link Local address
+        ///   </para>
+        /// </devdoc>
+        public bool IsIPv6LinkLocal
+        {
+            get
+            {
+                return (_family == AddressFamily.InterNetworkV6) &&
+                   ((_numbers[0] & 0xFFC0) == 0xFE80);
+            }
+        }
+
+        /// <devdoc>
+        ///   <para>
+        ///     Determines if an address is an IPv6 Site Local address
+        ///   </para>
+        /// </devdoc>
+        public bool IsIPv6SiteLocal
+        {
+            get
+            {
+                return (_family == AddressFamily.InterNetworkV6) &&
+                   ((_numbers[0] & 0xFFC0) == 0xFEC0);
+            }
+        }
+
+        public bool IsIPv6Teredo
+        {
+            get
+            {
+                return (_family == AddressFamily.InterNetworkV6) &&
+                       (_numbers[0] == 0x2001) &&
+                       (_numbers[1] == 0);
+            }
+        }
+
+        // 0:0:0:0:0:FFFF:x.x.x.x
+        public bool IsIPv4MappedToIPv6
+        {
+            get
+            {
+                if (AddressFamily != AddressFamily.InterNetworkV6)
+                {
+                    return false;
+                }
+                for (int i = 0; i < 5; i++)
+                {
+                    if (_numbers[i] != 0)
+                    {
+                        return false;
+                    }
+                }
+                return (_numbers[5] == 0xFFFF);
+            }
+        }
+
+        internal bool Equals(object comparandObj, bool compareScopeId)
+        {
+            IPAddress comparand = comparandObj as IPAddress;
+
+            if (comparand == null)
+            {
+                return false;
+            }
+
+            // Compare families before address representations
+            if (_family != comparand._family)
+            {
+                return false;
+            }
+            if (_family == AddressFamily.InterNetworkV6)
+            {
+                // For IPv6 addresses, we must compare the full 128-bit representation.
+                for (int i = 0; i < NumberOfLabels; i++)
+                {
+                    if (comparand._numbers[i] != _numbers[i])
+                    {
+                        return false;
+                    }
+                }
+
+                // The scope IDs must also match
+                return !compareScopeId || comparand._scopeId == _scopeId;
```

Comment:
@pgavlin Nit: (I'm very picky with this one): in the original code the first comparison was made to comparand._scopeId. To preserve the original operation order we should maybe change to `comparand._scopeId == _scopeId || !compareScopeId` ?